src/librustc_mir/interpret/visitor.rs

```diff
@@ -291,17 +292,28 @@ macro_rules! make_value_visitor {
                         // use that as an unambiguous signal for detecting primitives.  Make sure
                         // we did not miss any primitive.
                         debug_assert!(fields > 0);
-                        self.visit_union(v)?;
+                        self.visit_union(v)
                     },
                     layout::FieldPlacement::Arbitrary { ref offsets, .. } => {
-                        // FIXME: We collect in a vec because otherwise there are lifetime errors:
-                        // Projecting to a field needs (mutable!) access to `ecx`.
-                        let fields: Vec<EvalResult<'tcx, Self::V>> =
-                            (0..offsets.len()).map(|i| {
-                                v.project_field(self.ecx(), i as u64)
-                            })
-                            .collect();
-                        self.visit_aggregate(v, fields.into_iter())?;
+                        // Special handling needed for generators: All but the first field
+                        // (which is the state) are actually implicitly `MaybeUninit`, i.e.,
+                        // they may or may not be initialized, so we cannot visit them.
+                        match v.layout().ty.sty {
+                            ty::Generator(..) => {
```

Comment:
Niche code also has an exception for generator fields https://github.com/rust-lang/rust/blob/7a0cef74a8b859184b46d7ae6f7e01367fab5fb0/src/librustc/ty/layout.rs#L1812-L1817

Would it make sense to try to simplify all downstream code for generators by wrapping all its fields with `MaybeUninit` very early?

Reply:
Interesting, yes that would be the same exception.

I am not sure how complicated it would be for generators to do this wrapping.

Reply:
IMO generators should be treated like an `union` with field offsets.
Unless we want to generate "variants" for the states involved, which would be a bit more work, but would provide a safe view into the state of the generator.

Reply:
They don't have `Union` layout though, so right now they need special treatment everywhere.

Reply:
> The discriminant field is perfectly safe to read and we could even do value range restrictions on it to be able to use niche optimizations on generators.

In fact that would make perfect sense, it encodes the state after all and hence has a limited value range.